packages/datadog-instrumentations/src/http/server.js

```diff
@@ -23,10 +25,38 @@ addHook({ name: 'https' }, http => {
 
 addHook({ name: 'http' }, http => {
   shimmer.wrap(http.ServerResponse.prototype, 'emit', wrapResponseEmit)
+  shimmer.wrap(http.ServerResponse.prototype, 'end', wrapWriteHead)
+  shimmer.wrap(http.ServerResponse.prototype, 'writeHead', wrapWriteHead)
+  shimmer.wrap(http.ServerResponse.prototype, 'write', wrapWriteHead)
   shimmer.wrap(http.Server.prototype, 'emit', wrapEmit)
   return http
 })
 
+function wrapWriteHead (writeHead) {
+  return function (statusCode) {
+    if (this.finished) return
+
+    if (requestEndedSet.has(this)) {
+      return writeHead.apply(this, arguments)
+    }
+
+    requestEndedSet.add(this)
+
+    const abortController = new AbortController()
+
+    // TODO: this doesn't support headers sent with res.writeHead()
+    const responseHeaders = this.getHeaders()
+
```

Comment:
since headers can be either provided gradually and retrieved with `getHeaders()` or they can be passed to `writeHead()` as a positional argument, you're gonna need to merge those somehow